File: Python/ByRound/1580/1580_A_Portal.py
# Problem: CF 1580 A - Portal
# https://codeforces.com/contest/1580/problem/A

"""
Algorithm: Rectangle Inspection with Prefix Sums
Time Complexity: O(n^2 * m^2) where n and m are dimensions of the input matrix
Space Complexity: O(n * m) for prefix sum arrays

This solution finds the minimum number of operations to form a valid "portal" 
in the given rectangle. A portal is defined by specific conditions on borders 
and interior. The approach uses prefix sums to quickly query the number of 
operations needed to create a valid portal of a certain size. It iterates over 
all possible top-left and bottom-right coordinates of a 5x4 or larger rectangle 
and computes the required changes using prefix sums.

The function `v` calculates the number of 0s or 1s in a subrectangle using 
prefix sums, depending on the flag. It is used to count the number of 
changes needed in different parts of the candidate portal.
"""

import sys

T = int(sys.stdin.readline().strip())


def v(is0, is1, x, y, a, b, flag):
    # Function to calculate sum of values in a subrectangle using prefix sums
    # flag = 1 means count 1s, otherwise count 0s
    if flag == 1:
        g = is1
    else:
        g = is0
    return g[a + 1][b + 1] + g[x][y] - g[a + 1][y] - g[x][b + 1]


for t in range(T):
    n, m = list(map(int, sys.stdin.readline().strip().split(" ")))
    mat = []
    for i in range(n):
        mat.append(sys.stdin.readline().strip())
    
    # Create prefix sum arrays for 0s and 1s
    is1 = [[0 for j in range(m + 1)] for i in range(n + 1)]
    is0 = [[0 for j in range(m + 1)] for i in range(n + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            # Build prefix sum array for 1s
            is1[i][j] = (
                is1[i - 1][j]
                + is1[i][j - 1]
                - is1[i - 1][j - 1]
                + int(mat[i - 1][j - 1] == "1")
            )
            # Build prefix sum array for 0s
            is0[i][j] = (
                is0[i - 1][j]
                + is0[i][j - 1]
                - is0[i - 1][j - 1]
                + int(mat[i - 1][j - 1] == "0")
            )
    
    res = 16  # Initialize with a large value
    
    # Try all combinations of top-left and bottom-right coordinates
    for i in range(n):
        for j in range(m):
            # Iterate over possible bottom rows (a)
            for a in range(i + 4, n):
                # Calculate cost for left border (vertical edges)
                left = v(is0, is1, i + 1, j, a - 1, j, 0)
                if left >= res:
                    break  # Early termination if not promising
                
                # Iterate over possible right columns (b)
                for b in range(j + 3, m):
                    # Cost for top border (horizontal edges)
                    up = v(is0, is1, i, j + 1, i, b - 1, 0)
                    r = left + up
                    if r >= res:
                        break  # Early termination
                    
                    # Cost for bottom border (horizontal edges)
                    down = v(is0, is1, a, j + 1, a, b - 1, 0)
                    r += down
                    if r >= res:
                        break  # Early termination
                    
                    # Cost for interior (filling middle part)
                    mid = v(is0, is1, i + 1, j + 1, a - 1, b - 1, 1)
                    r += mid
                    if r >= res:
                        break  # Early termination
                    
                    # Cost for right border (vertical edges)
                    right = v(is0, is1, i + 1, b, a - 1, b, 0)
                    r += right
                    
                    # Update minimum operations if better result found
                    if r < res:
                        res = r
                        
    print(res)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/A/1580_A_Portal.cpp
// Problem: CF 1580 A - Portal
// https://codeforces.com/contest/1580/problem/A

/*
Algorithm: Portal
Purpose: Find the minimum number of operations to transform a given n x m grid into one that contains at least one valid portal sub-rectangle.
A portal is defined as an a x b rectangle where:
  - a >= 5, b >= 4
  - All edges except corners must be obsidian (1), and all inner cells must be empty (0)
  - Corners can be any type

Time Complexity: O(n^2 * m^2) where n,m <= 400
Space Complexity: O(n * m)

Approach:
- Precompute prefix sums for rows to efficiently check number of 1s in a range.
- Enumerate all possible top-left positions (i,j) of potential portals.
- For each such position, enumerate potential right edge k > j+2.
- Calculate cost to make boundary edges obsidian and inner part empty.
- Use precomputed prefix sums to count 0s in rows for efficiency.
*/
#include <algorithm>
#include <iostream>

using namespace std;
const int N = 409;
char s[N][N];
int p[N][N]; // Prefix sum of 1s in each row
int main() {
  ios::sync_with_stdio(0), cin.tie(0);
  int T, n, m, i, j, k, l, ans, now, tot;
  cin >> T;
  while (T--) {
    cin >> n >> m, ans = 20; // Initialize answer with high value
    for (i = 1; i <= n; ++i)
      cin >> (s[i] + 1);   // Read input grid starting from index 1
    // Precompute prefix sums for each row
    for (i = 1; i <= n; ++i)
      for (j = 1; j <= m; ++j)
        p[i][j] = p[i][j - 1] + (s[i][j] == '1');

    // Enumerate top-left corner (i,j) of a potential portal
    for (i = 1; i <= n; ++i)
      for (j = 1; j <= m; ++j) {
        tot = 0;
        // Extend right edge to k, k > j+2
        for (k = j + 1; k < m; ++k) {
          tot += (s[i][k] == '0');   // Count how many 0s we need to change
          if (tot >= ans)            // Early termination if cost exceeds current best
            break;
          if (k >= j + 2) {           // At least width 4
            now = tot;               // Initialize with changes to right edge
            // Enumerate bottom edge from l = i+1
            for (l = i + 1; l < n; ++l) {
              // Add new row's contribution to the cost of making edges obsidian
              now +=
                  p[l][k] - p[l][j] + (s[l][j] == '0') + (s[l][k + 1] == '0');
              if (now >= ans)
                break;
              // When we have at least 5 rows, check for valid portal
              if (l >= i + 3)
                ans = min(ans, now + k - j - p[l + 1][k] + p[l + 1][j]);
            }
          }
        }
      }
    cout << ans << '\n';
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/